src/components/heartrate/Ppg.h

```diff
@@ -3,29 +3,80 @@
 #include <array>
 #include <cstddef>
 #include <cstdint>
-#include "components/heartrate/Biquad.h"
-#include "components/heartrate/Ptagc.h"
+#include "arduinoFFT.h"
+#include "InterpolationLib.h"
 
 namespace Pinetime {
   namespace Controllers {
     class Ppg {
     public:
       Ppg();
-      int8_t Preprocess(float spl);
-      float HeartRate();
-
-      void SetOffset(uint16_t i);
-      void Reset();
+      int8_t Preprocess(float hrs, uint32_t als);
+      int HeartRate();
+      void Reset(bool resetDaqBuffer);
 
     private:
-      std::array<int8_t, 200> data;
-      size_t dataIndex = 0;
-      float offset;
-      Biquad hpf;
-      Ptagc agc;
-      Biquad lpf;
+      // The sampling frequency (Hz) set by delay in HeartRateTask.cpp
+      static constexpr float sampleFreq = 10.0f;
+      // Daq dataLength: Must be power of 2
+      static const int dataLength = 64;
+      // The frequency resolution (Hz)
+      const float freqResolution = sampleFreq / dataLength;
+      // Number of samples before each analysis
+      // 0.8 second update rate at 10Hz
+      const int overlapWindow = 8;
+      // Maximum number of spectrum running averages
+      // Note: actual number of spectra averaged = spectralAvgMax + 1
+      const int spectralAvgMax = 2;
+      // Multiple Peaks above this threshold (% of max) are rejected
+      const float peakDetectionThreshold = 0.6f;
+      // Maximum peak width (bins) at threshold for valid peak.
+      const float maxPeakWidth = 2.5f;
+      // Metric for spectrum noise level.
+      const float signalToNoiseThreshold = 3.0f;
+      // Heart rate Region Of Interest begin (bins)
+      const int hrROIbegin = static_cast<int>((30.0f / 60.0f) / freqResolution + 0.5f);
+      // Heart rate Region Of Interest end (bins)
+      const int hrROIend = static_cast<int>((240.0f / 60.0f) / freqResolution + 0.5f);
+      // Minimum HR (Hz)
+      const float minHR = 40.0f / 60.0f;
+      // Maximum HR (Hz)
+      const float maxHR = 230.0f / 60.0f;
+      // Threshold for high DC level after filtering
+      const float dcThreshold = 0.5f;
+      // ALS detection factor
+      const float alsFactor = 2.0f;
+
+      arduinoFFT FFT;
+      // Raw ADC data
+      std::array<float, dataLength> dataHRS;
```

Comment:
Just FIY : the previous implementation was mostly based on a 200B array. This improved implementation uses 4 arrays of 64 floats, which is 4 *  64 * 4 = 1024B.

This memory is statically allocated on the global stack (since those arrays are members of Ppg that is a member of HeartRateTask which is declared in the global scope in main.cpp

Reply:
Thanks! In an upcoming commit I was able to shave 128 bytes off the dataHRS buffer by changing to u16.